hal/src/rtl872x/sleep_hal.cpp

```diff
@@ -428,21 +457,61 @@ class SleepClass {
             PSRAM_datawr = ((temp << 8) | (temp << 24)) | (temps | (temps << 16));
             PSRAM_CTRL_CA_Gen(psram_ca, (BIT11 | BIT0), PSRAM_LINEAR_TYPE, PSRAM_REG_SPACE, PSRAM_WRITE_TRANSACTION);
             PSRAM_CTRL_DPin_Reg(psram_ca, &PSRAM_datawr, PSRAM_WRITE_TRANSACTION);
+
+            // Disable memory access from CPU
+            // PSRAM_DEV->CSR |= BIT_PSRAM_MEM_IDLE;
+            // while(!(PSRAM_DEV->CSR & BIT_PSRAM_MEM_IDLE));
+
+            // Wait for half-sleep entry just in case
+            DelayUs(200);
         } else {
             temp = HAL_READ32(SYSTEM_CTRL_BASE_LP, REG_AON_LDO_CTRL1);
             temp &= ~(BIT_LDO_PSRAM_EN);
             HAL_WRITE32(SYSTEM_CTRL_BASE_LP, REG_AON_LDO_CTRL1, temp);	
         }
+        __ISB();
+        __DSB();
     }
     
-    __attribute__((section(".ram.sleep"), optimize("O0")))
+    __attribute__((section(".ram.sleep")))
     void resumePsram() {
-        /* dummy read to make psram exit half sleep mode */
-        uint32_t temp = HAL_READ32(PSRAM_BASE, 0);
-        (void)temp;
-        /* wait psram exit half sleep mode */
-        DelayUs(100);
-        DCache_Invalidate((uint32_t)PSRAM_BASE, 32);
+        // Disable memory access from CPU
+        PSRAM_DEV->CSR |= BIT_PSRAM_MEM_IDLE;
+        while(!(PSRAM_DEV->CSR & BIT_PSRAM_MEM_IDLE));
+        uint8_t PSRAM_CA[6];
+        // FIXME: No idea how to align addresses properly here to correctly read things out, reading beginning of PSRAM seems to work though
+        PSRAM_CTRL_CA_Gen(PSRAM_CA, 0, PSRAM_LINEAR_TYPE, PSRAM_MEM_SPACE, PSRAM_READ_TRANSACTION);
+        for (int i = 0; i < 10; i++) {
+            uint32_t temp = 0xaabbccdd;
+            PSRAM_DEV->DPDR = 0;
+            PSRAM_DEV->CMD_DPIN_NDGE = (PSRAM_CA[0] | (PSRAM_CA[2] << 8) | (PSRAM_CA[4] << 16));
+            PSRAM_DEV->CMD_DPIN = (PSRAM_CA[1] | (PSRAM_CA[3] << 8) | (PSRAM_CA[5] << 16));
+
+            PSRAM_DEV->CSR &= (~BIT_PSRAM_DPIN_MODE);
+            PSRAM_DEV->CSR |= PSRAM_DPIN_READ_MODE;
+
+            PSRAM_DEV->CCR = BIT_PSRAM_DPIN;
+            // 150us should be enough, but we'll anyway retry a few times just in case
+            for (int j = 0; j < 150; j++) {
+                if (PSRAM_DEV->CCR & BIT_PSRAM_DPIN) {
+                    break;
+                }
+                DelayUs(1);
```

Comment:
`DelayUs()` Is a ROM function right? 
Do we know if it is safe to call when attempting to re-enable PSRAM?